### PR TITLE
Use structured suggestion for disambiguating method calls

### LINKED ISSUE #65635: Incorrect compiler hint for method with (&self) and (&mut self) signature when multiple applicable items are found  in scope
Code: 
```rust
trait A {
    fn method(&self) {
        println!("A");
    }
}

trait B {
    fn method(&self) {
        println!("B");
    }
}

struct Empty {}

impl A for Empty {}
impl B for Empty {}

fn main() {
    Empty {}.method();
}
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c870592075712e461a1a83545bd86e38)

Error: 
```
error[E0034]: multiple applicable items in scope
  --> src/main.rs:19:14
   |
19 |     Empty {}.method();
   |              ^^^^^^ multiple `method` found
   |
note: candidate #1 is defined in an impl of the trait `A` for the type `Empty`
  --> src/main.rs:2:5
   |
2  |     fn method(&self) {
   |     ^^^^^^^^^^^^^^^^
   = help: to disambiguate the method call, write `A::method(Empty {})` instead
note: candidate #2 is defined in an impl of the trait `B` for the type `Empty`
  --> src/main.rs:8:5
   |
8  |     fn method(&self) {
   |     ^^^^^^^^^^^^^^^^
   = help: to disambiguate the method call, write `B::method(Empty {})` instead
```
Should be `A::method(&Empty {})` and `B::method(&Empty {})`.

### PR DESCRIPTION
**Pull request by estebank**
Fix #65635.

### CODE REVIEW

File: src/test/ui/error-codes/E0034.stderr [hunk 1]
```diff
@@ -9,13 +9,19 @@ note: candidate #1 is defined in an impl of the trait `Trait1` for the type `Tes
...
-   = help: to disambiguate the method call, write `Trait1::foo(...)` instead
 note: candidate #2 is defined in an impl of the trait `Trait2` for the type `Test`
   --> $DIR/E0034.rs:16:5
    |
 LL |     fn foo() {}
    |     ^^^^^^^^
-   = help: to disambiguate the method call, write `Trait2::foo(...)` instead
+help: disambiguate the method call for candidate #1
+   |
+LL |     Trait1::foo(...)()
```
Comment (Centril) [inline src/test/ui/error-codes/E0034.stderr]: This looks strange?

File: src/librustc_typeck/check/method/suggest.rs [hunk 2]
```diff
@@ -82,34 +82,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let print_disambiguation_help = |
```
Comment (Centril) [inline src/librustc_typeck/check/method/suggest.rs]: Time to move this closure out to a proper method due to size?

File: src/librustc_typeck/check/method/suggest.rs [hunk 3]
```diff
@@ -82,34 +82,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-            err.help(&format!(
-                "to disambiguate the method call, write `{}::{}({}{})` instead",
-                trait_name,
-                item_name,
-                if rcvr_ty.is_region_ptr() && args.is_some() {
-                    if rcvr_ty.is_mutable_ptr() {
-                        "&mut "
+            let mut applicability = Applicability::MachineApplicable;
+            let sugg_args = if let ty::AssocKind::Method = kind {
+                format!(
```
Comment (Centril) [inline src/librustc_typeck/check/method/suggest.rs]: In particular this branch could be moved.

File: src/librustc_typeck/check/method/suggest.rs (hunk 2 above)
Comment (estebank) [reply src/librustc_typeck/check/method/suggest.rs]: done

File: src/test/ui/error-codes/E0034.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/error-codes/E0034.stderr]: fixed

File: src/test/ui/issues/issue-18446.stderr [hunk 4]
```diff
@@ -2,7 +2,10 @@ error[E0034]: multiple applicable items in scope
   --> $DIR/issue-18446.rs:18:7
    |
 LL |     x.foo();
-   |       ^^^ multiple `foo` found
+   |     --^^^--
+   |     | |
+   |     | multiple `foo` found
+   |     help: disambiguate the method call for candidate #2: `T::foo(&x)`
```
Comment (varkor) [inline src/test/ui/issues/issue-18446.stderr]: Why is there only one disambiguation suggestion here?

Comment (estebank) [reply src/test/ui/issues/issue-18446.stderr]: The code for this is

```

trait T {
    fn foo(&self);
}

impl<'a> dyn T + 'a {
    fn foo(&self) {}
}

impl T for i32 {
    fn foo(&self) {}
}

fn main() {
    let x: &dyn T = &0i32;
    x.foo(); //~ ERROR multiple applicable items in scope [E0034]
}
```

I'm not sure what the appropriate syntax would be to call `&dyn T::foo`. Also note that we were _already_ not suggesting it.

Comment (varkor) [reply src/test/ui/issues/issue-18446.stderr]: > Also note that we were already not suggesting it.

I did notice this. I think we ought to open a new issue — if the user can't specify `(&dyn T)::foo`, we should at least explain this in the diagnostic. But this PR looks good as it is.

**Pull request merged**